core/src/main/java/org/elasticsearch/common/text/UTF8SortedAsUnicodeComparator.java

```diff
@@ -18,60 +18,76 @@
  */
 package org.elasticsearch.common.text;
 
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefIterator;
 import org.elasticsearch.common.bytes.BytesReference;
 
+import java.io.IOException;
 import java.util.Comparator;
 
 // LUCENE 4 UPGRADE: Is this the right way of comparing bytesreferences inside Text instances?
 // Copied from Lucene's BytesRef comparator
-public class UTF8SortedAsUnicodeComparator implements Comparator<BytesReference> {
-
+public final class UTF8SortedAsUnicodeComparator implements Comparator<BytesReference> {
     public final static Comparator<BytesReference> utf8SortedAsUnicodeSortOrder = new UTF8SortedAsUnicodeComparator();
-
     // Only singleton
-    private UTF8SortedAsUnicodeComparator() {
-    }
-
+    private UTF8SortedAsUnicodeComparator() {}
     @Override
     public int compare(BytesReference a, BytesReference b) {
-        if (a.hasArray() && b.hasArray()) {
-            final byte[] aBytes = a.array();
-            int aUpto = a.arrayOffset();
-            final byte[] bBytes = b.array();
-            int bUpto = b.arrayOffset();
-
-            final int aStop = aUpto + Math.min(a.length(), b.length());
-            while (aUpto < aStop) {
-                int aByte = aBytes[aUpto++] & 0xff;
-                int bByte = bBytes[bUpto++] & 0xff;
-
-                int diff = aByte - bByte;
-                if (diff != 0) {
-                    return diff;
-                }
-            }
+        try {
+            // we use the iterators since it's a 0-copy comparison where possible!
+            final BytesRefIterator aIter = a.iterator();
+            final BytesRefIterator bIter = b.iterator();
+            final long lengthToCompare = Math.min(a.length(), b.length());
+            BytesRef aRef = aIter.next();
+            BytesRef bRef = bIter.next();
+            if (aRef != null && bRef != null) { // do we have any data?
+                if (aRef.length == a.length() && bRef.length == b.length()) { // is it only one array slice we are comparing?
+                    return aRef.compareTo(bRef);
+                } else {
+                    for (int i = 0; i < lengthToCompare;) {
+                        if (aRef.length == 0) {
+                            aRef = aIter.next();
+                            if (aRef == null) {
```

Comment:
wouldn't it be a bug if this happens? it would mean the iterator returns null even though less than `lengthToCompare` bytes have been compared so far, which is supposed to be less than the length of the ref